File: Question_LeetcodeAndGeeks/Greedy/medium/Q010Leetcode731_my-calendar-ii/Q010Leetcode731.cpp
// #include <bits/stdc++.h>
#include <iostream>
#include <vector>
#include <string>

#define vi vector<int>
#define vii vector<vi>

using namespace std;

static const auto magic = []() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    return nullptr;
}();

class MyCalendarTwo
{
public:
    map<int, int> bst;
    MyCalendarTwo()
    {
        bst.clear();
    }

    bool book(int start, int end)
    {
        bst[start]++;
        bst[end]--;

        int count = 0;
        for (auto &key : bst)
        {
            count += key.second;
            if (count > 2)
            {
                bst[start]--;
                bst[end]++;
                if (bst[start] == 0)
                    bst.erase(start);
                if (bst[end] == 0)
                    bst.erase(end);

                return false;
            }
        }
        return true;
    }
};

File: Question_LeetcodeAndGeeks/Greedy/medium/Q010Leetcode731_my-calendar-ii/Q010Leetcode731.java
import java.util.TreeMap;
import java.util.Map;
public class Q010Leetcode731 {
    class MyCalendarTwo {

        TreeMap < Integer, Integer > map = new TreeMap < > ();
        public MyCalendarTwo() {
            map.clear();
        }

        public boolean book(int start, int end) {
            map.put(start, map.getOrDefault(start, 0) + 1);
            map.put(end, map.getOrDefault(end, 0) - 1);

            int count = 0;
            for (Map.Entry < Integer, Integer > key: map.entrySet()) {
                count += key.getValue();
                if (count > 2) {
                    map.put(start, map.getOrDefault(start, 0) - 1);
                    map.put(end, map.getOrDefault(end, 0) + 1);
                    if (map.get(start) == 0) map.remove(start);
                    if (map.get(end) == 0) map.remove(end);
                    return false;
                }
            }
            return true;
        }
    }
}